From 8de479c079048c2e5c346b69eef0b156b9cf5e73 Mon Sep 17 00:00:00 2001
From: Alex Marginean <alexandru.marginean@freescale.com>
Date: Mon, 30 Oct 2017 13:50:32 +0200
Subject: [PATCH 085/706] enetc: added spoof checking support

Prevents VFs from sending out traffic with SMAC != station MAC(s).
It can be configured using:
ip link set eth0|eth1 vf 0|1 spoofchk on|off

Signed-off-by: Alex Marginean <alexandru.marginean@freescale.com>
(cherry picked from commit fc483ffce2a2e3e9698c9273cedf798f040d2ea7)
Signed-off-by: Zumeng Chen <zumeng.chen@windriver.com>
---
 drivers/net/ethernet/freescale/enetc/enetc.c    | 16 ++++++++++++++++
 drivers/net/ethernet/freescale/enetc/enetc_hw.h |  1 +
 2 files changed, 17 insertions(+)

diff --git a/drivers/net/ethernet/freescale/enetc/enetc.c b/drivers/net/ethernet/freescale/enetc/enetc.c
index bbb59eb5c94a..25b615559980 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc.c
+++ b/drivers/net/ethernet/freescale/enetc/enetc.c
@@ -1288,6 +1288,21 @@ static int enetc_set_vf_vlan(struct net_device *ndev, int vf, u16 vlan,
 	return 0;
 }
 
+static int enetc_set_vf_spoofchk(struct net_device *ndev, int vf, bool setting)
+{
+	struct enetc_ndev_priv *priv = netdev_priv(ndev);
+	u32 cfgr;
+
+	if (vf > priv->si->num_vfs)
+		return -EINVAL;
+
+	cfgr = enetc_port_rd(&priv->si->hw, ENETC_PV0CFGR(vf + 1));
+	cfgr = (cfgr & ~ENETC_PVCFGR_ASE) | (setting ? ENETC_PVCFGR_ASE : 0);
+	enetc_port_wr(&priv->si->hw, ENETC_PV0CFGR(vf + 1), cfgr);
+
+	return 0;
+}
+
 static const struct net_device_ops enetc_ndev_ops = {
 	.ndo_open		= enetc_open,
 	.ndo_stop		= enetc_close,
@@ -1297,6 +1312,7 @@ static const struct net_device_ops enetc_ndev_ops = {
 	.ndo_set_rx_mode	= enetc_set_rx_mode,
 	.ndo_set_vf_mac		= enetc_set_vf_mac,
 	.ndo_set_vf_vlan	= enetc_set_vf_vlan,
+	.ndo_set_vf_spoofchk	= enetc_set_vf_spoofchk,
 };
 
 static const struct net_device_ops enetc_ndev_vf_ops = {
diff --git a/drivers/net/ethernet/freescale/enetc/enetc_hw.h b/drivers/net/ethernet/freescale/enetc/enetc_hw.h
index 03d1535bd5f3..ba8e5d86a1a5 100644
--- a/drivers/net/ethernet/freescale/enetc/enetc_hw.h
+++ b/drivers/net/ethernet/freescale/enetc/enetc_hw.h
@@ -101,6 +101,7 @@ enum enetc_bdr_type {TX, RX};
 #define ENETC_PV0CFGR(n)	(0x00920 + (n) * 0x10)  /* n = SI index */
 #define ENETC_PVCFGR_SET_TXBDR(val)	((val) & 0xff)
 #define ENETC_PVCFGR_SET_RXBDR(val)	(((val) & 0xff) << 16)
+#define ENETC_PVCFGR_ASE	BIT(15)
 
 #define ENETC_RSSHASH_KEY_SIZE	40
 #define ENETC_PRSSK(n)		(0x01410 + (n) * 4) /* n = [0..9] */
-- 
2.17.1

